### ISSUE TITLE
jj shows changes as hidden, even some that have visible child changes

### ISSUE THREAD

**Issue by colemickens** [martinvonz/jj#3861]
## Description

Something seems to be off with my repo, or my current jj's understanding of my repo.

I have this chunk of commits: https://gist.github.com/colemickens/63fe33ae402fc110475be831c1a8cbca

As you can see, some of the hidden ones have visible descendent changes. I understand this to be impossible.

I can't say when this started happening. I guess it's not the end of the world, but I did want to rebase one of these hidden commits, and I can't do that while jj thinks its hidden.

I did just snapshot the entire repository's directory, checkout, git, jj, etc. I'd rather not share if I don't have to, but if it proves to be useful, I can.

cc: @ilyagr who pointed out that the visible children was a sign of a possible bug. Thanks @ilyagr and jj contributors.

## Steps to Reproduce the Problem

1. Unclear.

## Expected Behavior

jj presents an internally consistent view of the repository history.

## Actual Behavior

it doesn't. :(

## Specifications

- Platform: linux
- Version: 2d0b715

**Comment by colemickens** [comment]
Maybe this is related?

```
╭ slynux  ~/code/nixcfg shell-ci-env 11min 49sec 14ms colemickens/yoga7x main* ✘1
╰🡒  fd --hidden --no-ignore-vcs 'sync-conflict'
.git/index.sync-conflict-20240519-190533-M7QB72N
.git/index.sync-conflict-20240524-091850-M7QB72N
.jj/working_copy/checkout.sync-conflict-20240519-190533-M7QB72N
.jj/working_copy/tree_state.sync-conflict-20240519-190533-M7QB72N
.jj/working_copy/tree_state.sync-conflict-20240605-203617-M7QB72N
```

The timestamps do roughly line up...

**Comment by yuja** [comment]
If you have some background process like rsync, you can get this kind of inconsistency in the git backend layer.
`jj debug reindex` might help solve the immediate problem.

**Comment by ilyagr** [comment]
For some more context, see https://martinvonz.github.io/jj/latest/technical/concurrency/#syncing-with-rsync-nfs-dropbox-etc.

It'd be great to fix such bugs and to make such problems fewer, so thanks for testing out the workflow. However, you should also expect some problems. They are unlikely to lead to data loss, though.

I think you shared enough to conclude that you encountered a syncing-related concurrency bug, but not enough to debug it further. Yuya is the expert on these kinds of problems; I'm not sure if this is a known problem (some of which we don't have a good way to address) or a new one. If you shared the repo, we could track down the exact operation that caused the problem using `jj op log` and `jj log --at-op`, and the exact states of the repo before and after that; that's the main thing we'd gain.

**Comment by colemickens** [comment]
> jj debug reindex

beautiful! fantastic! seems to have fixed it right up!

@ilyagr I think if I understand correctly, jj is meant to auto-recover in these cases? Or maybe recover is not the right word, but something. I'll have to wait for later to be able to pull the copy I made, but I can send it if you think it will help narrow things down.

**Comment by ilyagr** [comment]
I am not actually sure *what* `jj` is supposed to do in these cases, it could be tricky. I merely hope that something better is possible.

**Comment by yuja** [comment]
Closed as dup of #2193
